### PR TITLE
Fix eos_token problem in all required models

### LINKED ISSUE #1478: [Qwen2 tokenizer] Ensure ``eos_token`` is not added if ``add_eos=False``
Fix similar to #1477 

Add a test and ensure it passes.

### LINKED ISSUE #1479: [Mistral tokenizer] Ensure ``eos_token`` is not added if ``add_eos=False``
Fix similar to #1477

### LINKED ISSUE #1480: [Gemma tokenizer] Ensure ``eos_token`` is not added if ``add_eos=False``
Fix similar to #1477

### LINKED ISSUE #1481: [Phi3 tokenizer] Ensure ``eos_token`` is not added if ``add_eos=False``
Fix similar to #1477

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [X] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
Closes #1478 
Closes #1479 
Closes #1480 
Closes #1481 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [X] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [X] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [X] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [X] I have added an example to docs or docstrings

So generally for Mistral and Gemma - just check to pass None if add_eos = False in `tokenize_messages_no_special_tokens`
For Qwen and Phi - same check as in #1477 in truncate(actually for Phi should be reviewed in more accurate way)

And 4 unittests for all models.

### CODE REVIEW

**Comment by krammnic** [comment]
Actually, I think that tests with this big lists of token ids should be refactored(maybe some fixture?).

**Comment by krammnic** [comment]
@RdoubleA @joecummings Require review

File: torchtune/models/phi3/_tokenizer.py [hunk 1]
```diff
@@ -101,13 +101,11 @@ def encode(
             trim_leading_whitespace=trim_leading_whitespace,
         )
 
-    def decode(self, ids: List[int], skip_special_tokens: bool = True) -> str:
```
Comment (joecummings) [inline torchtune/models/phi3/_tokenizer.py:104]: Why did you remove `skip_special_tokens`?

Comment (krammnic) [reply torchtune/models/phi3/_tokenizer.py:104]: Weird, maybe accidentally removed. Let me fix

**Comment by krammnic** [comment]
> Actually, I think that tests with this big lists of token ids should be refactored(maybe some fixture?).

Will think about it more and maybe open PR

**Comment by krammnic** [comment]
Something really weird.

File: torchtune/models/phi3/_tokenizer.py (hunk 1 above)
Comment (krammnic) [reply torchtune/models/phi3/_tokenizer.py]: Fixed

**Review by joecummings — APPROVED** [review]
LGTM
